crawl-ref/source/dgn-overview.cc

```diff
@@ -1094,3 +1064,164 @@ bool connected_branch_can_exist(branch_type br)
+        // (D) Dungeon        (T) Temple         (L) Lair           etc.
+        // at most 4 branches on 1 line
+        clear_messages();
+        int linec = 0;
+        string line;
+        for (branch_type br : brs)
+        {
+            if (linec == 4)
+            {
+                linec = 0;
+                mpr(line);
+                line = "";
+            }
+            line += make_stringf("(%c) %-14s ",
+                                 branches[br].travel_shortcut,
+                                 branches[br].shortname);
+            ++linec;
+        }
+        if (!line.empty())
+            mpr(line);
+
+        // tells you some special keys, but not all
+        mprf(MSGCH_PROMPT, "Annotate which branch? (. - %s, ? -help)",
+             lid.describe(false, true).c_str());
+
+        int keyin = get_ch();
+        switch (keyin)
+        {
+        CASE_ESCAPE
+            return ID_CANCEL;
+        case '?':
+            return ID_HELP;
+        case '\n': case '\r': case '.':
+            return ID_HERE;
+        case '<':
+            return ID_UP;
+        case '>':
+            return ID_DOWN;
+        case CONTROL('P'):
+            {
+                const branch_type parent = parent_branch(lid.branch);
+                if (parent < NUM_BRANCHES)
+                    return parent;
+            }
+            break;
+        default:
+            // Is this a branch hotkey?
+            for (branch_type br : brs)
+            {
+                if (toupper_safe(keyin) == branches[br].travel_shortcut)
+                    return br;
+            }
+            // Otherwise cancel
+            return ID_CANCEL;
+        }
+    }
+}
+
+void do_annotate()
+{
+    level_id lid  = level_id::current();
+    int branch = _prompt_annotate_branch(lid);
+    int depth;
+
+    if (branch == ID_HELP) // XXX: this help message could be improved.
+    {
+        clear_messages();
+        mpr("press the letter corresponding to the branch of the level you want to annotate, or");
+        mpr("? for help, . or enter to annotate the level you're currently in, < and > to annotate");
+        mpr("the levels upstairs and downstairs respectively, and control-P for the parent branch");
+        do_annotate();
+        return;
+    }
+
+    if (branch == ID_CANCEL)
+    {
+        canned_msg(MSG_OK);
+        return;
+    }
+
+    if (branch == ID_HERE)
+    {
+        annotate_level(lid);
+        return;
+    }
+
+    if (branch == ID_UP)
+    {
+        lid = find_up_level(lid);
+        annotate_level(lid);
+        return;
+    }
+
+    if (branch == ID_DOWN)
+    {
+        lid = find_down_level(lid);
+        annotate_level(lid);
+        return;
+    }
+
+    int max_depth = branches[branch].numlevels;
+    // Handle one-level branches by not prompting.
+    if (max_depth==1)
+        depth = 1;
+
+    clear_messages();
+    const string prompt = make_stringf ("What level of %s? ", branches[branch].longname);
```

Comment:
no whitespace after `make_stringf`

Reply:
I tried it, and we really do need that space